sphinx_gallery/py_source_parser.py

```diff
@@ -19,30 +19,57 @@
 """
 
 
-def get_docstring_and_rest(filename):
-    """Separate `filename` content between docstring and the rest
+def parse_file_source(filename):
+    """Parse source file into AST node
+
+    Parameters
+    ----------
+    filename : str
+        File path
 
-    Strongly inspired from ast.get_docstring.
 
     Returns
     -------
-    docstring: str
-        docstring of `filename`
-    rest: str
-        `filename` content without the docstring
+    node : AST node
+    content : utf-8 encoded string
     """
+
     # can't use codecs.open(filename, 'r', 'utf-8') here b/c ast doesn't
-    # seem to work with unicode strings in Python2.7
-    # "SyntaxError: encoding declaration in Unicode string"
+    # work with unicode strings in Python2.7 "SyntaxError: encoding
+    # declaration in Unicode string" In python 2.7 the string can't be
+    # encoded and have information about its encoding. That is particularly
+    # problematic since source files include in their header information
+    # about the file encoding.
+    # Minimal example to fail: ast.parse(u'# -*- coding: utf-8 -*-')
+
     with open(filename, 'rb') as fid:
         content = fid.read()
     # change from Windows format to UNIX for uniformity
     content = content.replace(b'\r\n', b'\n')
 
     try:
         node = ast.parse(content)
+        return node, content.decode('utf-8')
     except SyntaxError:
-        return SYNTAX_ERROR_DOCSTRING, content.decode('utf-8'), 1
+        return SyntaxError, content.decode('utf-8')
```

Comment:
This feels a bit kludgy to return an exception class rather than a node. Why don't you SyntaxError through and catch it in `get_docstring_and_rest` and have an except clause at the `parse_file_source` call?

Reply:
I agree is not the nicest way. I ended up using this, because I need this function in few more places and I rely those places always having access to the decoded content and to know if the parsing was successful or not.

Reply:
I saw that but I would rather let the SyntaxError go through and catch it where you call `get_docstring_and_rest`. Arguably you could have a SyntaxError somewhere else in your code but I don't think this is that bad. If we care about that we could catch the SyntaxError in `parse_source_file` and raise a custom exception e.g. ParsingError (or a better name) then catch ParsingError where parse_source_file` is called.

Reply:
This is on the boundaries of my knowledge. But if I raise the exception how do I get access to the returned value of content?

Reply:
Just in case I am missing and the exception approach does not work, can you not return None instead?

Reply:
Of course. I can return None